packages/@aws-cdk/toolkit-lib/lib/api/cloud-assembly/cached-source.ts

```diff
@@ -0,0 +1,50 @@
+import type * as cxapi from '@aws-cdk/cx-api';
+import { BorrowedAssembly } from './private/borrowed-assembly';
+import type { ICloudAssemblySource, IReadableCloudAssembly } from './types';
+
+/**
+ * A CloudAssemblySource that is caching its result once produced.
+ *
+ * Most Toolkit interactions should use a cached source. Not caching is
+ * relevant when the source changes frequently and it is to expensive to predict
+ * if the source has changed.
+ *
+ * The `CachedCloudAssembly` is both itself a readable CloudAssembly, as well as
+ * a Cloud Assembly Source. The lifetimes of cloud assemblies produced by this
+ * source are coupled to the lifetime of the `CachedCloudAssembly`. In other
+ * words: the `dispose()` functions of those cloud assemblies don't do anything;
+ * only the `dispose()` function of the `CachedCloudAssembly` will be used.
+ *
+ * FIXME: We should consider referencing counting here, although that seems
```

Comment:
What improvement would this bring? Might be good to spell it out here, in case someone else picks this up in the future.